File: apps/app3/src/main/java/pt/obs/kafka/AppKafkaConsumer.java
package pt.obs.kafka;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.Metrics;
import lombok.extern.slf4j.Slf4j;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;
import pt.obs.db.DataEntity;
import pt.obs.db.DataRepository;

@Slf4j
@Component
public class AppKafkaConsumer {

    private Counter topicCounter;
    private final DataRepository dataRepository;
    private String restOutUrl;

    AppKafkaConsumer(DataRepository dataRepository,
                     @Value("${metrics.counter.topic-in.name}") String topicCounterName,
                     @Value("${rest.out.url}") String restOutUrl) {
        this.dataRepository = dataRepository;
        this.topicCounter = Metrics.counter(topicCounterName, "it-1", "it-2");
        this.restOutUrl = restOutUrl;
    }

    @KafkaListener(id="app3:1", topics = {"topic1", "topic3"})
    void listen(ConsumerRecord<String, String> record){
        log.info(STR."Fetch data from topic \{record.topic()}: \{record.key()}=\{record.value()}");
        topicCounter.increment();

        DataEntity data = new DataEntity();
        data.setData(STR."AppKafkaConsumer: \{topicCounter.count()}");
        log.info(STR."Write data to database: \{data}");
        dataRepository.save(data);

        log.info(STR."Call REST URL: \{restOutUrl}, result: \{RestClient.create(restOutUrl).get().retrieve().toEntity(String.class)}");
    }
}

File: apps/app4/internal/kafka/AppKafkaConsumer.go
package kafka

import (
	"context"
	"fmt"
	"github.com/segmentio/kafka-go"
	"net/http"
	"os"
	"pt.observability.elastic/app4/internal/db"
	log "pt.observability.elastic/app4/internal/logging"
	"pt.observability.elastic/app4/internal/metrics"
)

var in_topics = []string{"topic1", "topic2", "topic3"}

var topicCounter = metrics.NewCounter(
	"topic_in_counter",
	map[string]string{"it_1": "it-2"})

func Listen(ctx context.Context) {
	for _, topic := range in_topics {
		go listenOnTopic(ctx, topic)
	}
}

func listenOnTopic(ctx context.Context, topic string) {
	address := os.Getenv("KAFKA_BOOTSTRAP-SERVERS")

	if address == "" {
		address = "localhost:9092"
	}

	restOutUrl := os.Getenv("REST_OUT_URL")

	if restOutUrl == "" {
		restOutUrl = "http://localhost:8081/api-3"
	}
	reader := kafka.NewReader(kafka.ReaderConfig{
		Brokers:   []string{address},
		Topic:     topic,
		Partition: 0,
		MaxBytes:  10e6, // 10MB
	})
	log.Info(ctx, "Listen on topic: ", topic)

	for {
		message, err := reader.FetchMessage(ctx)

		if err != nil {
			log.Error(ctx, err)
			break
		}
		ctx, span := startSpan(ctx, &message)
		log.Info(ctx, fmt.Sprintf("Fetch data from topic %s: %s=%s", topic, message.Key, message.Value))
		topicCounter.Increment(ctx)

		save(ctx, fmt.Sprintf("AppKafkaConsumer: %d", topicCounter.Count()))

		response, _ := http.Get(restOutUrl)
		log.Info(ctx, fmt.Sprintf("Call REST URL: %s, result: %s", restOutUrl, response))
		span.End()
	}

	if err := reader.Close(); err != nil {
		log.Error(ctx, "failed to close reader:", err)
	}
	log.Info(ctx, "close reader: "+topic)
}

func save(ctx context.Context, data string) {
	dataEntity := db.DataEntity{Data: data}
	log.Info(ctx, fmt.Sprintf("Write data to database: %s", dataEntity))
	db.Save(ctx, dataEntity)
}
